src/clandestine.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019. The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+
+//! Functionality for securely handling sensitive data in a clandestine way.
+//! Some data, like cryptographic key material, needs special handling.
+//! This kind of data should be zeroized when it goes away, should never be written to log files or displayed, and
+//! should be generally difficult to misuse. This library tackles all three.
+//!
+//! We store the data on the heap and use `zeroize` to ensure it is cleared when it goes out of scope.
+//! We use a `Hidden` wrapper to prevent the data from being displayed or written to debug logs.
+//! We use strict type separation to prevent the data from being used out of context, which is especially important for
+//! key material. And we only allow the data to be accessed by reference and mutable reference, which prevents
+//! unintended moves or copies. You can still clone, but this retains all the guardrails and is safe.
+
+use std::{fmt::Display, marker::PhantomData};
+
+use zeroize::Zeroize;
+
+use crate::Hidden;
+
+/// A marker trait for differentiated clandestine types
+pub trait ClandestineType {}
```

Comment:
I would rename it to `ClandestineMarker`.

Reply:
That's not Rustic convention.

Reply:
> I would rename it to `ClandestineMarker`.

That's not Rust convention. which is one of the code smalls that says this isn't quite the right approach here.

Reply:
A better name might be something like `ClandestineLabel`. The idea is that I can define some type like `type KeyForPurposeA = Clandestine<[u8; 32], clandestine!(LabelForPurposeA)>`, and another type like `type KeyForPurposeB = Clandestine<[u8; 32], clandestine!(LabelForPurposeB)>`. Even though the underlying type `[u8; 32]` is the same in both, keys of one type can't be confused with keys of the other type since the compiler will get angry.

The actual marker trait itself isn't strictly necessary, but it seemed like a good way to enforce a common derivation for context labels. I wasn't sure of a better overall approach that reduced boilerplate while still meeting the goals.

Reply:
Updated for clarity.